base/char.jl

```diff
@@ -29,11 +45,9 @@ done(c::Char, state) = state
 isempty(c::Char) = false
 in(x::Char, y::Char) = x == y
 
-==(x::Char, y::Char) = UInt32(x) == UInt32(y)
-isless(x::Char, y::Char) = UInt32(x) < UInt32(y)
-
-const hashchar_seed = 0xd4d64234
-hash(x::Char, h::UInt) = hash_uint64(((UInt64(x)+hashchar_seed)<<32) ⊻ UInt64(h))
+==(x::Char, y::Char) = reinterpret(UInt32, x) == reinterpret(UInt32, y)
+isless(x::Char, y::Char) = reinterpret(UInt32, x) < reinterpret(UInt32, y)
+hash(x::Char, h::UInt) = hash(reinterpret(UInt32, x), hash(Char, h))
```

Comment:
According to `BenchmarkTools` this is 5 times slower than the old `hash(x::Char)`.  Should we use 
```jl
hash(x::Char, h::UInt) = hash_uint64(((reinterpret(UInt32, x)+UInt64(hashchar_seed))<<32) ⊻ UInt64(h))
```
or similar instead?

Reply:
Perhaps. I worry about that xor being symmetric. I also worry that we're using this kind of pattern all over Base and it is quite inefficient. We really need a better way to express this.

Reply:
Went with your suggested definition for now. If there's some hash collision issue with the existing definition, that's a totally independent issue to this representation change so can be addressed separately.